flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -0,0 +1,302 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Map;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.EnvironmentSettings;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.test.util.MiniClusterWithClientResource;
+import org.apache.flink.types.Row;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.flink.source.BoundedTableFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestFlinkUpsert extends FlinkCatalogTestBase {
+
+  @ClassRule
+  public static final MiniClusterWithClientResource MINI_CLUSTER_RESOURCE =
+      MiniClusterResource.createWithClassloaderCheckDisabled();
+
+  @ClassRule
+  public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+  private final boolean isStreamingJob;
+  private final Map<String, String> tableUpsertProps = Maps.newHashMap();
+  private TableEnvironment tEnv;
+
+  public TestFlinkUpsert(String catalogName, Namespace baseNamespace, FileFormat format, Boolean isStreamingJob) {
+    super(catalogName, baseNamespace);
+    this.isStreamingJob = isStreamingJob;
+    tableUpsertProps.put(TableProperties.FORMAT_VERSION, "2");
+    tableUpsertProps.put(TableProperties.UPSERT_ENABLED, "true");
+    tableUpsertProps.put(TableProperties.DEFAULT_FILE_FORMAT, format.name());
+  }
+
+  @Parameterized.Parameters(name = "catalogName={0}, baseNamespace={1}, format={2}, isStreaming={3}")
+  public static Iterable<Object[]> parameters() {
+    List<Object[]> parameters = Lists.newArrayList();
+    for (FileFormat format : new FileFormat[] {FileFormat.PARQUET, FileFormat.AVRO, FileFormat.ORC}) {
+      for (Boolean isStreaming : new Boolean[] {true, false}) {
+        for (Object[] catalogParams : FlinkCatalogTestBase.parameters()) {
```

Comment:
Probably we only need to test this for `hive` and `hadoop` catalog. Realistically, even just one would be fine.

I'd actually probaby rather add `write.distribution-mode` to the tests, as there was a bug fix where Flink itself didn't shuffle data on equality fields (primary key fields) in some upsert situations. It was backported to Flink 1.14.

Reply:
Agreed. We don't need different catalogs since this is the same path for all of them. And good call on the distribution mode!

Reply:
Do we reasonably need to test Parquet, Avro, and ORC? I'd say just Parquet. The only difference is that Avro doesn't have stats and would have incorrectly passed tests.

Reply:
Agreed. hadoop should suffice.
We can just test Parquet.

Reply:
Yeah agreed. It's been beneficial while testing locally.

And admittedly the issue did not occur with Avro originally due to the lack of stats as mentioned, so I'd be fine with just parquet and one catalog.

I'd even prefer to test all 3 and only one catalog, because the test is fairly end to end. But I'm fine with only parquet and then two catalogs.

Reply:
On further thought, I think we should test at least ORC and Parquet.

Although there is a logic bug in this case, these newly added end to end tests rely heavily on statistics and metrics stats, which is directly linked to file format.

I'd prefer to use one catalog and all 3 file formats, or at least ORC and parquet.

Reply:
+1 for testing one catalog and all 3 file formats.  Catalog won't introduce any difference for this change in my thought, but file formats may do. (I saw the failure OOM unit tests seems were all ORC tests).

Reply:
Updated this to use hadoop catalog only.

There doesn't seem to be an equivalent enum to `SparkCatalogConfig` that allows me to pass just one catalog easily, so I've just hardcoded the name `testhadoop` into the test.

We can refactor that later if we want.